eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/StringFormattedExceptionMessageCheck.java

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FullIdent;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Checks that the message provided to "logger.throwableAt*().log" is static (not created using String.format).
+ */
+public class StringFormattedExceptionMessageCheck extends AbstractCheck {
+    static final String ERROR_MESSAGE = "Short message passed to \"logger.throwableAt*().log\" should be static. "
+        + "Provide dynamic components using the \"addKeyValue(key, value)\" method instead. "
+        + "See https://github.com/Azure/azure-sdk-for-java/wiki/Client-core:-logging-exceptions-best-practices for more details.";
+    private static final Set<String> THROWABLE_AT_LOGGING_METHODS = new HashSet<>(Arrays.asList(
+        ".throwableAtError",
+        ".throwableAtWarning"));
+
+    private static final String LOG_METHOD_NAME = ".log";
+    private static final String STRING_FORMAT_METHOD_NAME = "String.format";
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.METHOD_CALL,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        if (token.getType() == TokenTypes.METHOD_CALL) {
+            String name = FullIdent.createFullIdentBelow(token).getText();
+            if (THROWABLE_AT_LOGGING_METHODS.stream().anyMatch(name::endsWith)) {
+                DetailAST logMethodCall = token.getParent().getParent();
+                if (logMethodCall == null || logMethodCall.getType() != TokenTypes.METHOD_CALL) {
+                    return;
+                }
+
+                String nextName = FullIdent.createFullIdentBelow(logMethodCall).getText();
+                if (nextName.endsWith(LOG_METHOD_NAME)) {
+                    DetailAST elist = logMethodCall.findFirstToken(TokenTypes.ELIST);
+                    if (elist == null) {
+                        return;
+                    }
+
+                    DetailAST logExpr = elist.findFirstToken(TokenTypes.EXPR);
+                    if (logExpr == null) {
+                        return;
+                    }
+
+                    String logFirstArgMethod = FullIdent.createFullIdent(logExpr.getFirstChild().getFirstChild()).getText();
+                    if (logFirstArgMethod.endsWith(STRING_FORMAT_METHOD_NAME)) {
```

Comment:
Should we also check for string concatenation like `"Retry count = " + count` too?

Reply:
thanks for the suggestion, I think I found a reasonable heuristic to flag - a combination of string literal and identifier.